File: DSA/0084-largest-rectangle-in-histogram/solution.cpp
class Solution {
public:
    vector<int> p(vector<int>& h){
        stack<int>st;
        vector<int>ans(h.size());
        for(int i=0;i<h.size();++i){
            while(!st.empty() && h[st.top()]>=h[i]){
                st.pop();
            }
            if(st.empty()) ans[i]=-1; //-1 
            else ans[i]=st.top(); // -1 -1 1
            st.push(i);// 1 // 5
        }
        return ans;
    }
    vector<int> n(vector<int>& h){
        stack<int>st;
        vector<int>ans(h.size());
        for(int i=h.size()-1;i>=0;--i){
            while(!st.empty() && h[st.top()]>h[i]){
                st.pop();
            }
            if(st.empty()) ans[i]=-1; //-1 
            else ans[i]=st.top(); // -1 -1 1
            st.push(i);// 1 // 5
        }
        return ans;
    }
    int largestRectangleArea(vector<int>& h) {
        vector<int>pre=p(h);
        vector<int>nex=n(h);
         int ans=INT_MIN;
         for(int i=0;i<h.size();++i){
            if(nex[i]==-1) nex[i]=h.size();
            int area=nex[i]-pre[i]-1;
            int total=h[i]*area;
            ans=max(ans,total);
         }
         return ans;
    }
};

File: DSA/0084-largest-rectangle-in-histogram/solution.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int size=heights.length;
        int[] left=new int[size];
        Stack<Integer> st=new Stack();
        for(int i=0;i<size;i++){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i]){
                st.pop();
            }
            left[i]=st.isEmpty() ? -1 : st.peek();
            st.push(i);
        }

        st.clear();

        int[] right=new int[size];
        for(int i=size-1;i>=0;i--){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i]){
                st.pop();
            }
            right[i]=st.isEmpty() ? size : st.peek();
            st.push(i);
        }

        int ans=Integer.MIN_VALUE;
        for(int i=0;i<size;i++){
            int width=right[i]-left[i]-1;
            ans=Math.max(ans,heights[i]*width);

        }
        return ans;

    }
}
